### PR TITLE
[CI, enhancement] add pytest-cov for testing

### PR DESCRIPTION
**Pull request by icfaust**
## Description

Add the possibility to run pytest-cov for test coverage reports for CI, and add them to the Github Actions CI for PRs.  There will be a follow-up PR if we chose to integrate this into coveralls.io or codecov based on internal discussions.

How to use these files: 

- navigate to your GitHub Actions CI run (for example: https://github.com/intel/scikit-learn-intelex/actions/runs/11851114955)
- Download and extract all files from all artifacts
- Install coverage.py :https://coverage.readthedocs.io/en/7.6.5/
- run: ```coverage combine /coverage_lnx_Py3.10_1.2 /coverage_lnx_Py3.9_1.1 /coverage_lnx_Py3.11_1.3 /coverage_win_Py3.10_1.2 /coverage_win_Py3.9_1.1 /coverage_win_Py3.11_1.3``` https://coverage.readthedocs.io/en/latest/cmd.html#combining-data-files-coverage-combine
- Generate report of your choice example: ```coverage report``` https://coverage.readthedocs.io/en/latest/cmd.html#reporting
---

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [x] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [x] I have provided justification why performance has changed or why changes are not expected.
- [x] I have provided justification why quality metrics have changed or why changes are not expected.
- [x] I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

### CODE REVIEW

**Review by samir-nasibli — COMMENTED** [review]
Thank you @icfaust for that important tool integration.
I will take a look today later.

File: .ci/scripts/run_sklearn_tests.py
Comment (ahuber21) [inline .ci/scripts/run_sklearn_tests.py:41]:
```suggestion
    sklearn_file_dir = os.path.dirname(sklearn.__file__)

```

File: .ci/scripts/run_sklearn_tests.py
Comment (ahuber21) [inline .ci/scripts/run_sklearn_tests.py:57]:
```suggestion
    if rc := os.getenv("COVERAGE_RCFILE"):
        pytest_args += (
            f'--cov=onedal --cov=sklearnex --cov-config={rc} '
            "--cov-report=term".split(" ")
        )

```

Since we no longer support python < 3.8 we can start using the walrus operator.

File: .ci/scripts/run_sklearn_tests.py
Comment (ahuber21) [inline .ci/scripts/run_sklearn_tests.py:55]: Why was the conversion added? What were non-int returns previously?

File: onedal/basic_statistics/tests/test_basic_statistics.py
Comment (ahuber21) [inline onedal/basic_statistics/tests/test_basic_statistics.py:1]: Please provide an explanation for these changes

File: .ci/scripts/run_sklearn_tests.py
Comment (icfaust) [reply .ci/scripts/run_sklearn_tests.py:55]: Conversion to int occurs at the end of the file, I wanted to have it earlier to check for a successful run, so I moved it into the return code.

File: onedal/basic_statistics/tests/test_basic_statistics.py
Comment (icfaust) [reply onedal/basic_statistics/tests/test_basic_statistics.py:1]: Good question, it was the only test file getting picked up for code coverage because of this inclusion, which other testing files are dependent on.  I simplified it and moved it into a location which was more conducive for wider use (onedal.tests.utils).  Let me know what you think.

**Comment by david-cortes-intel** [comment]
@icfaust Would be nice to get an HTML report in addition to plain text.

File: .github/workflows/ci.yml
Comment (david-cortes-intel) [inline .github/workflows/ci.yml:207]: If this is for windows, shouldn't the quote start before the name of the variable that you are setting?

File: requirements-test.txt
Comment (david-cortes-intel) [inline requirements-test.txt:4]: Last version is 6.0 - why use an older one?

File: .ci/scripts/run_sklearn_tests.py
Comment (david-cortes-intel) [inline .ci/scripts/run_sklearn_tests.py:56]:
```suggestion
            "--cov=onedal", "--cov=sklearnex", f"--cov-config={rc}",
            "--cov-report=term",

```

File: conda-recipe/run_test.sh
Comment (david-cortes-intel) [inline conda-recipe/run_test.sh:82]: I think without the quotes it'd end up failing if there spaces in the path.

File: .github/workflows/ci.yml
Comment (david-cortes-intel) [inline .github/workflows/ci.yml:215]: We're using `NO_DIST=1` elsewhere.

**Comment by samir-nasibli** [comment]
> ## Description
> Add the possibility to run pytest-cov for test coverage reports for CI, and add them to the Github Actions CI for PRs.
> 
> Checklist to comply with **before moving PR from draft**:
> 
> **PR completeness and readability**
> 
> * [x]  I have reviewed my changes thoroughly before submitting this pull request.
> * [x]  I have commented my code, particularly in hard-to-understand areas.
> * [x]  I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
> * [x]  Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
> * [x]  I have added a respective label(s) to PR if I have a permission for that.
> * [x]  I have resolved any merge conflicts that might occur with the base branch.
> 
> **Testing**
> 
> * [x]  I have run it locally and tested the changes extensively.
> * [x]  All CI jobs are green or I have provided justification why they aren't.
> * [x]  I have extended testing suite if new functionality was introduced in this PR.
> 
> **Performance**
> 
> * [x]  I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
> * [x]  I have provided justification why performance has changed or why changes are not expected.
> * [x]  I have provided justification why quality metrics have changed or why changes are not expected.
> * [x]  I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

Could you please add to the description how to use this?
If this already integrated into CI, could you please share link for statistics?

File: requirements-test.txt
Comment (icfaust) [reply requirements-test.txt:4]: Good catch, the version 6.0 was released as this PR was waiting for review. Will update.

File: .github/workflows/ci.yml
Comment (icfaust) [reply .github/workflows/ci.yml:216]: Good catch, luckily in this case the quotation marks do not impact the operation of the underlying check. Will swap it nevertheless

File: conda-recipe/run_test.sh
Comment (icfaust) [reply conda-recipe/run_test.sh:82]: Oh shoot, that was definitely an oversight. I'll make sure the quotes are readded.

File: .github/workflows/ci.yml
Comment (icfaust) [reply .github/workflows/ci.yml]: changed.

File: .ci/scripts/run_sklearn_tests.py
Comment (icfaust) [reply .ci/scripts/run_sklearn_tests.py]: removed term as well since its assumed that the reports will be viewed only after integration together.

**Comment by icfaust** [comment]
> > ## Description
> > Add the possibility to run pytest-cov for test coverage reports for CI, and add them to the Github Actions CI for PRs.
> > Checklist to comply with **before moving PR from draft**:
> > **PR completeness and readability**
> > 
> > * [x]   I have reviewed my changes thoroughly before submitting this pull request.
> > * [x]   I have commented my code, particularly in hard-to-understand areas.
> > * [x]   I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
> > * [x]   Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
> > * [x]   I have added a respective label(s) to PR if I have a permission for that.
> > * [x]   I have resolved any merge conflicts that might occur with the base branch.
> > 
> > **Testing**
> > 
> > * [x]   I have run it locally and tested the changes extensively.
> > * [x]   All CI jobs are green or I have provided justification why they aren't.
> > * [x]   I have extended testing suite if new functionality was introduced in this PR.
> > 
> > **Performance**
> > 
> > * [x]   I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
> > * [x]   I have provided justification why performance has changed or why changes are not expected.
> > * [x]   I have provided justification why quality metrics have changed or why changes are not expected.
> > * [x]   I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.
> 
> Could you please add to the description how to use this? If this already integrated into CI, could you please share link for statistics?

I'll add to the description, but as of now all it will do is store the necessary artifacts for use locally on your dev machine or via a service like https://about.codecov.io/ or https://coveralls.io/  That will come after necessary internal discussion of which direction to go.

File: .github/workflows/ci.yml
Comment (icfaust) [reply .github/workflows/ci.yml:216]: Turns out changing this interprets the ```|``` into an operation and will fail. Reverted to original.

File: .github/workflows/ci.yml
Comment (icfaust) [reply .github/workflows/ci.yml:215]: Turns out it broke github actions. Had to revert. Why? Because exporting to GITHUB_ENV via a batch file seems very poorly documented/ operates in an unexpected manner.  Would do it in linux, but I need the full path of the coverage rc file from windows, and running a bash script from github actions forces it into a linux environment (another odd behavior).

Comment (david-cortes-intel) [reply .github/workflows/ci.yml:215]: But why would a '1' in this case make such a difference? Or was it about a different change?

File: sklearnex/linear_model/logistic_regression.py
Comment (Vika-F) [inline sklearnex/linear_model/logistic_regression.py:292]: What is the purpose of those comments? And why are they added only for LogReg?

File: onedal/tests/utils/__init__.py
Comment (Vika-F) [inline onedal/tests/utils/__init__.py:27]: Is it possible to place this definition into `basic_statistics/tests` module?
Because `utils` does not seem to be a proper place for this kind of definitions.

I like the overall basic statistics tests refactoring, it's just the placement of this dictionary looks not so good to me.

File: sklearnex/linear_model/logistic_regression.py
Comment (icfaust) [reply sklearnex/linear_model/logistic_regression.py:292]: Good question! We can avoid code coverage analysis for regions between comments ```coverage: gpu_start``` and ```coverage: gpu_stop``` since we do not do any GPU operations in public CI. If someone does a full GPU-only PR they will see a large decrease in code coverage, which we can designate as GPU code.  Do to so, we need to remove the ```#``` from this file: https://github.com/icfaust/scikit-learn-intelex/blob/dev/pytest-cov/.coveragerc#L23

Comment (icfaust) [reply sklearnex/linear_model/logistic_regression.py:292]: Oh, and I did it only for LogisticRegression as an initial thing because the sklearnex code is only GPU-relevant, as CPU is offloaded to daal4py (and would otherwise show very bad code-coverage metrics).

File: onedal/tests/utils/__init__.py
Comment (icfaust) [reply onedal/tests/utils/__init__.py:27]: I'll move it an add another omission to the coverage.rc file for that case, otherwise it will be picked up in coverage metrics.

File: .github/workflows/ci.yml
Comment (icfaust) [reply .github/workflows/ci.yml:215]: Something with how github is writing and reading to the GITHUB_ENV file is dropping the 1 when done from a batch script. Quotations or making it a string with "" will keep the quotations too. Valid strings for the input can be found here: https://github.com/intel/scikit-learn-intelex/blob/main/setup.py#L85 so it should be okay.

File: sklearnex/linear_model/logistic_regression.py
Comment (Vika-F) [reply sklearnex/linear_model/logistic_regression.py:292]: Thanks for the explanation.
But it looks like a lot of work to write all those comments properly in the code and maintain them in good quality.
Can there be a better way to exclude GPU-related code from the code coverage analysis?

Comment (Vika-F) [reply sklearnex/linear_model/logistic_regression.py:292]: Maybe it is possible to add something like:
```
exclude_also =
    self\._onedal_estimator\..*queue=queue\)
```
to .coveragerc. To exclude all the estimators calls that use SYCL queue.

File: .coveragerc
Comment (icfaust) [reply .coveragerc:16]: Done! Will address the other issue tomorrow.

File: sklearnex/linear_model/logistic_regression.py
Comment (icfaust) [reply sklearnex/linear_model/logistic_regression.py:292]: Since we aren't making a hard and fast rule, it could be that we make the first round of analysis without gpu exclusion, the metrics will show regressions with GPU code introduction, but the way GPU code isn't fully standardized in the codebase, so you are right that it would be a long a tedious process.

Comment (icfaust) [reply sklearnex/linear_model/logistic_regression.py:292]: I'll start by removing the secondary excludes from LogisticRegression and elsewhere, and a development ticket to the backlog (let me know what you think @Vika-F ).  I think this will occur after we integrate into coveralls.

**Review by ethanglaser — COMMENTED** [review]
Its looking good - can you remind what the conclusion of how this info would be used? Like would we have a check that sees if coverage worsens with changes or just manual analysis for now?

**Comment by ethanglaser** [comment]
> Is there an example of manually generated report for this?

From "Archive coverage data" step in public CI there is https://github.com/intel/scikit-learn-intelex/actions/runs/11838651610/artifacts/2187948371 for example, not sure the best way to view the contents though

File: sklearnex/basic_statistics/tests/test_incremental_basic_statistics.py
Comment (icfaust) [reply sklearnex/basic_statistics/tests/test_incremental_basic_statistics.py:1]: The coverage metrics were picking up this test file for coverage because of how functionality was being re-used, it was moved to a separate file so that test coverage would not be generated for test functions.

File: .github/workflows/ci.yml
Comment (icfaust) [reply .github/workflows/ci.yml:291]: There will be two coverage output files following coverage.py's convention (see this example: https://coverage.readthedocs.io/en/latest/cmd.html#combining-data-files-coverage-combine). This is not done yet in this PR, but all the various coverage files will be combined and then can be used to generate reports: https://coverage.readthedocs.io/en/latest/cmd.html#reporting.  Ideally we would integrate into https://coveralls.io/ (what IntelPython uses) or https://about.codecov.io/ (what sklearn uses) which will automatically calculate coverage values and will add comments in PRs.  We could create HTML files for seeing which lines of a PR are unused, but public CI doesn't contain any GPU code testing, and as of now GitHub Actions CI doesn't test our supported sklearn versions.  So long story short: which way would you @ethanglaser  like to take it? Ideally I would want to do it in a follow-up PR.

File: .ci/scripts/run_sklearn_tests.py
Comment (icfaust) [reply .ci/scripts/run_sklearn_tests.py]: removed

File: onedal/tests/utils/__init__.py
Comment (icfaust) [reply onedal/tests/utils/__init__.py:1]: removed from the code, reverted to previous.

File: conda-recipe/run_test.bat
Comment (icfaust) [reply conda-recipe/run_test.bat]: done, also changed ``` python``` above it

Comment (david-cortes-intel) [reply conda-recipe/run_test.bat]: I think the `defined` call requires having the variable name without the percent signs, otherwise it will apply substitutions on it and the statement might not work. Also looks like the double quotes were lost.

File: conda-recipe/run_test.bat
Comment (david-cortes-intel) [inline conda-recipe/run_test.bat:23]:
```suggestion
IF NOT DEFINED PYTHON (
```

File: conda-recipe/run_test.bat
Comment (david-cortes-intel) [inline conda-recipe/run_test.bat:31]:
```suggestion
IF DEFINED COVERAGE_RCFILE (set COV_ARGS=--cov=onedal --cov=sklearnex --cov-config="%COVERAGE_RCFILE%" --cov-append --cov-report=) else (set COV_ARGS=)
```

File: conda-recipe/run_test.bat
Comment (icfaust) [reply conda-recipe/run_test.bat]: Changes as requested made.

**Review by david-cortes-intel — APPROVED** [review]

**Review by Alexsandruss — CHANGES_REQUESTED** [review]

File: .coveragerc
Comment (icfaust) [reply .coveragerc:40]: @Alexsandruss done. please review

File: conda-recipe/run_test.sh
Comment (david-cortes-intel) [inline conda-recipe/run_test.sh:60]: I think this doesn't do the same thing. It would end up trying to execute a command instead of adding arguments to an existing call.

File: conda-recipe/run_test.sh
Comment (david-cortes-intel) [inline conda-recipe/run_test.sh:60]: I think this ends up adding only the first argument.

File: conda-recipe/run_test.sh
Comment (icfaust) [reply conda-recipe/run_test.sh]: good catch, we should schedule a meeting or make a chat to iron out these issues quickly. i have no particular preference

File: conda-recipe/run_test.sh
Comment (david-cortes-intel) [inline conda-recipe/run_test.sh:60]:
```suggestion
    printf -- "${ARGS[*]}"
```

File: onedal/basic_statistics/tests/utils.py
Comment (ethanglaser) [inline onedal/basic_statistics/tests/utils.py:1]: This is much cleaner than what we have in main branch. Nice.

File: .github/workflows/ci.yml
Comment (ethanglaser) [reply .github/workflows/ci.yml:291]: All sounds good - it would be interesting to get something set up with python team to see how they are using this

**Review by ethanglaser — APPROVED** [review]
No objections from my side assuming private CI is good and comments around bash/bat/yml syntax will be resolved

File: conda-recipe/run_test.sh
Comment (icfaust) [reply conda-recipe/run_test.sh:60]: Looks like the change isn't needed. It produced the requisite coverage files

Comment (david-cortes-intel) [reply conda-recipe/run_test.sh:60]: Does it still produce them if you add both the json report argument and the coverage argument?

Comment (icfaust) [reply conda-recipe/run_test.sh]: @david-cortes-intel I checked the file and you are right that it is incomplete, I have changed it to ```"${ARGS[*]}"``` along with some other necessary changes. I will not merge without your approval because of the changes to the report generation lines.

File: conda-recipe/run_test.bat
Comment (david-cortes-intel) [inline conda-recipe/run_test.bat:37]: Actually on second thought, I think these double quotes around `%COVERAGE_RCFILE%` would have an unintended effect of leaving quotes in the argument, so perhaps it should either be done without the quotes, or with the quotes around the whole `set "var=value"` call.

Sorry for the confusion.

File: conda-recipe/run_test.sh
Comment (icfaust) [reply conda-recipe/run_test.sh]: @david-cortes-intel I have added a matching test to the run_test.sh to match run_test.bat for checking file generation. I hope that will help address the file generation in linux question.

File: conda-recipe/run_test.bat
Comment (david-cortes-intel) [inline conda-recipe/run_test.bat:47]: Maybe it could print the `PYTEST_ARGS` template here.

File: conda-recipe/run_test.bat
Comment (david-cortes-intel) [inline conda-recipe/run_test.bat:49]: I'm not sure here if having a space before `||` makes a difference or not, but I think it'd be better to leave it just in case, just for clarity that it's not part of the argument.

**Review by david-cortes-intel — APPROVED** [review]
LGTM, just an ammendment about my earlier comment around the double quotes in the .bat file.

File: conda-recipe/run_test.bat
Comment (icfaust) [reply conda-recipe/run_test.bat:49]: Sounds good, I will let CI continue running to verify things as they are and will make changes combined with whatever failures pop up.

File: conda-recipe/run_test.bat
Comment (icfaust) [reply conda-recipe/run_test.bat:37]: No worries, have made each of the PYTEST_ARGS in quotes as you've suggested (on next commit)

File: conda-recipe/run_test.bat
Comment (icfaust) [reply conda-recipe/run_test.bat:48]: Added!

**Pull request merged**
